### ISSUE TITLE
solve_linear_system & linsolve returning None & EmptySet when there exists solution(s)

### ISSUE THREAD

**Issue by asadullah-7** [sympy/sympy#26223]
for the below system (in augmented form):
[[0, 3.00000000000000, 5], [0, -1, -1.6666666666666667]]

here y=5/3 and x is a free variable

so ideally the functions should return {sym2: 5/3}

I thought it was a precision issue, and tried everything I could like using Decimal, Fraction, S, and even used my Gaussian elimination with back substitution algorithm which I made some time back to check and it worked fine (although the precision was abs_tol of 1e-08 in this) and it worked fine.

since np.linalg.solve only works for systems with unique solution, I made a change in the 2nd equation (row) such that x=1
[[0, 3.00000000000000, 5], [1, -1, -0.6666666666666667]]

and numpy worked, linsolve worked, so did solve_linear_system.

PS: the system of equations is not directly in my control, it a result of processing other user inputs, though sympy is the only library used so conversion of sympy objects to other objects and vice versa isn't the issue here [most probably].

Sympy 1.12 on M2 mac

**Comment by oscarbenjamin** [comment]
Please show actual code rather than using words to describe what you are doing.

A system like this can only be solved with exact arithmetic:
```
In [2]: M = Matrix([[0, 3.00000000000000, 5], [0, -1, -1.6666666666666667]])

In [3]: linsolve(M)
Out[3]: ∅

In [5]: M.applyfunc(nsimplify)
Out[5]: 
⎡0  3    5  ⎤
⎢           ⎥
⎣0  -1  -5/3⎦

In [6]: linsolve(M.applyfunc(nsimplify))
Out[6]: {(τ₀, 5/3)}
```
The system has no solutions unless the bottom row is an exact multiple of the top row which it is not for the floats that you showed.

**Comment by asadullah-7** [comment]
> The system has no solutions unless the bottom row is an exact multiple of the top row which it is not for the floats that you showed.
that's my question. they are multiple of each other.

even with python floats:
`In [1]: -5/3.00000000000000==-1.6666666666666667
Out[1]: True`

> Please show actual code rather than using words to describe what you are doing.
sorry, I'm new to this.

**Comment by oscarbenjamin** [comment]
> they are multiple of each other.

They are not exact multiples and even if they were there is no guarantee that a floating point algorithm will find that after other rounding errors.

This kind of system of redundant equations where the lhs matrix does not have full rank cannot be solved in floating point reliably. There are other things that can be done like finding a least squares solution. You can also just remove the redundant rows:
```
In [22]: M = Matrix([[0, 3.00000000000000, 5], [0, -1, -1.6666666666666667]])

In [23]: M.rref()[0]
Out[23]: 
⎡0  1  1.66666666666667⎤
⎢                      ⎥
⎣0  0         0        ⎦   # <-- there is no guarantee to get exact zeros here

In [24]: M.rref()[0][:-1,:]
Out[24]: [0  1  1.66666666666667]

In [25]: linsolve(M.rref()[0][:-1,:])
Out[25]: {(τ₀, 1.66666666666667)}
```
Probably though the best thing to do here is to use rational numbers since these numbers are clearly intended to be rational numbers even if they are provided here as floats. I have shown how to convert the floats to rational. Another question though is why you have this system of redundant equations at all.

**Comment by pantheraleo-7** [comment]
> Another question though is why you have this system of redundant equations at all.

I forgot to tell, the program is about solving ODEs (I'm aware of dsolve). In IVPs or BVPs we could have these kind of systems of redundant equations. The system is not a direct input, rather an intermediary result.

> This kind of system of redundant equations where the lhs matrix does not have full rank cannot be solved in floating point reliably.

ok. it is kinda strange though that, for this case, rref succeeded in eliminating the redundant equation while linsolve couldn't or else it would have returned the correct solution. It is strange because linsolve uses Gauss-Jordan elimination which is basically transforming the given matrix into its rref! and solve_linear_system uses Gaussian elimination with back substitution which should be even more reliable. 

> Probably though the best thing to do here is to use rational numbers

basically, we need to sanitise the inputs to linsolve (or solve_linear_system)

**Comment by oscarbenjamin** [comment]
> which should be even more reliable

None of these things can be reliable if you are using floating point with a system such as this. There is no robust way to determine the rank of a matrix if the values in the matrix are not exact and/or if the arithmetic with those values is not exact.

**Comment by pantheraleo-7** [comment]
well it turns out solve() worked. so never mind. I didn't use solve at first because I thought why to use a general purpose function when we have specialised, "better" functions available for the job.

it is really a matter of worry that, for a given problem, some functions work while some don't. sympy is lacking consistency at the moment.

**Comment by oscarbenjamin** [comment]
> well it turns out solve() worked.

That is because solve turns the floats into rationals using nsimplify like I showed above. That is not always guaranteed to work though. You really need to use exact rational numbers from the start of the calculation and throughout if you want to be able to solve something like this reliably.

**Comment by pantheraleo-7** [comment]
> tried everything I could like using Decimal, Fraction, S

^as I did at the start. but now when you said it, I used sympy's Rational

`Rational(-5, 3)`
`-5/3`

this worked (with linsolve & solve_linear_system)

`Rational(-5/3)`
`-7505999378950827/4503599627370496`

this didn't

I know -5/3 is converted into a python float first, and then passed to the function. but Rational should be able to convert it no? the output it gives is not beautiful but it is correct, kind of

`-7505999378950827/4503599627370496 == -5/3`
`True`

or does sympy works differently than python here?



also, one more query, why did this system failed:
`[[0, 3.00000000000000, 5], [0, -1, -5/3]]`

because of 3.00000000000000?
